### PR TITLE
*: add a flag `RestoreWithDefaultDB` for Restore

### PR DESCRIPTION
**Pull request by rebelice**
### What problem does this PR solve? 
It's a part of [tidb-pr](https://github.com/pingcap/tidb/pull/21275)

### What is changed and how it works?

- add a flag `RestoreWithDefaultDB` for Restore
- modify some format case

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Review by lzmhhh123 — COMMENTED** [review]
LGTM

File: ast/dml.go [hunk 1]
```diff
@@ -983,6 +986,9 @@ func (n *SelectStmt) Restore(ctx *format.RestoreCtx) error {
 		if n.TableHints != nil && len(n.TableHints) != 0 {
 			ctx.WritePlain("/*+ ")
 			for i, tableHint := range n.TableHints {
+				if i != 0 {
+					ctx.WritePlain(", ")
+				}
```
Comment (kennytm) [inline ast/dml.go:991]: No, table hints *are* supposed to be separated by spaces, not commas. Demo on MySQL 8:

```
mysql> select /*+ set_var(sql_mode='x') set_var(timestamp=1) */ @@timestamp;
+-------------+
| @@timestamp |
+-------------+
|    1.000000 |
+-------------+
1 row in set, 1 warning (0.00 sec)

mysql> select /*+ set_var(sql_mode='x'), set_var(timestamp=1) */ @@timestamp;
+-------------------+
| @@timestamp       |
+-------------------+
| 1606280193.327941 |
+-------------------+
1 row in set, 2 warnings (0.00 sec)
```

**Review by kennytm — CHANGES_REQUESTED** [review]

File: ast/dml.go (hunk 1 above)
Comment (rebelice) [reply ast/dml.go:991]: But we use commas in our [user docs](https://docs.pingcap.com/zh/tidb/stable/optimizer-hints), so we need to modify the user docs?

**Comment by rebelice** [comment]
I have modified， PTAL @kennytm

File: ast/dml.go (hunk 1 above)
Comment (kennytm) [reply ast/dml.go]: better modify the docs to align with MySQL. the optional comma can be thought as TiDB's syntax extension.

File: format/format.go [hunk 2]
```diff
@@ -278,16 +282,25 @@ func (rf RestoreFlags) HasSpacesAroundBinaryOperationFlag() bool {
 	return rf.has(RestoreSpacesAroundBinaryOperation)
 }
 
+func (rf RestoreFlags) HasWithDefaultDB() bool {
+	return rf.has(RestoreWithDefaultDB)
```
Comment (kennytm) [inline format/format.go:286]: IMO returning `rf.DefaultDB != ""` is enough, no need to introduce a flag

File: format/format.go [hunk 3]
```diff
@@ -278,16 +282,25 @@ func (rf RestoreFlags) HasSpacesAroundBinaryOperationFlag() bool {
...
+}
+
+func (rf RestoreFlags) HasStringWithoutDefaultCharset() bool {
+	return rf.has(RestoreStringWithoutDefaultCharset)
+}
```
Comment (kennytm) [inline format/format.go:291]: this doesn't seem to have any effect

**Review by kennytm — COMMENTED** [review]
Rest LGTM

File: format/format.go (hunk 3 above)
Comment (rebelice) [reply format/format.go:291]: This is used in TiDB, related pr is [here](https://github.com/pingcap/tidb/pull/21275)

File: format/format.go (hunk 2 above)
Comment (rebelice) [reply format/format.go]: Done.

**Review by lzmhhh123 — APPROVED** [review]

**Comment by lzmhhh123** [comment]
The test failed due to an unstable test in the planner.

**Pull request merged**